### PR TITLE
use latest Buffer API instead of new Buffer

### PR DESCRIPTION
**Pull request by julianduque**
`new Buffer` has been deprecated in Node, this PR uses `Buffer.from` and
`Buffer.alloc` instead of `new Buffer` which are safer and performant.
Also provides an alternative to use the `safe-buffer` module for
backwards compatibility.

### CODE REVIEW

**Comment by kumavis** [comment]
i noticed this commit has diffs from your previous PR
in general diffs should only pertain to the PR at hand

**Comment by julianduque** [comment]
@kumavis yes, I think is the whitespace fixes, I can update the PR's

**Comment by julianduque** [comment]
Rebased with master

File: lib/hooked.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
+const Buffer = require('safe-buffer').Buffer
 const inherits = require('util').inherits
 const async = require('async')
 const ethUtil = require('ethereumjs-util')
 const Account = require('ethereumjs-account')
 const FakeMerklePatriciaTree = require('fake-merkle-patricia-tree')
 const VM = require('./index.js')
-const ZERO_BUFFER = new Buffer('0000000000000000000000000000000000000000000000000000000000000000', 'hex')
+const ZERO_BUFFER = Buffer.from('0000000000000000000000000000000000000000000000000000000000000000', 'hex')
```
Comment (axic) [inline lib/hooked.js]: Couldn't you use `Buffer.alloc(32)` here?

File: lib/index.js [hunk 2]
```diff
@@ -46,7 +47,7 @@ function VM (opts = {}) {
 
   if (this.opts.activatePrecompiles) {
     for (var i = 1; i <= 4; i++) {
-      this.trie.put(new Buffer('000000000000000000000000000000000000000' + i, 'hex'), new Account().serialize())
+      this.trie.put(Buffer.from('000000000000000000000000000000000000000' + i, 'hex'), new Account().serialize())
```
Comment (axic) [inline lib/index.js]: Maybe `Buffer.from(new BN(i).toArray('be', 32))` is clear of intent? Speed doesn't matter much here.

File: lib/opFns.js [hunk 3]
```diff
@@ -124,16 +125,16 @@ module.exports = {
     if (!exponent.isZero()) {
       var bytes = 1 + logTable(exponent)
       subGas(runState, new BN(bytes).muln(fees.expByteGas.v))
-      result = new Buffer(base.redPow(exponent).toArray('be', 32))
+      result = Buffer.from(base.redPow(exponent).toArray('be', 32))
     } else {
-      result = new Buffer([1])
+      result = Buffer.from([1])
```
Comment (axic) [inline lib/opFns.js]: You've used `Buffer.alloc(1)` elsewhere. Can we settle on a single format?

File: package.json [hunk 4]
```diff
@@ -12,10 +12,12 @@
     "ethereumjs-util": "4.5.0",
     "fake-merkle-patricia-tree": "^1.0.1",
     "functional-red-black-tree": "^1.0.1",
-    "merkle-patricia-tree": "^2.1.2"
+    "merkle-patricia-tree": "^2.1.2",
+    "safe-buffer": "^5.1.1"
   },
   "devDependencies": {
     "babel-preset-es2015": "^6.24.0",
+    "babel-preset-stage-0": "^6.24.1",
```
Comment (axic) [inline package.json]: What is this transformation doing?

File: lib/opFns.js [hunk 5]
```diff
@@ -43,15 +44,15 @@ module.exports = {
...
+    return Buffer.from(r)
   },
   SDIV: function (a, b, runState) {
     a = utils.fromSigned(a)
     b = utils.fromSigned(b)
 
     var r
     if (b.isZero()) {
-      r = new Buffer([0])
+      r = Buffer.alloc(1)
```
Comment (axic) [inline lib/opFns.js]: Thinking a bit more about this it may make the code more readable if not using `alloc(1)`, but rather `Buffer.from([0])` in `opFns.js`.

File: lib/hooked.js (hunk 1 above)
Comment (fanatid) [inline lib/hooked.js]: `Buffer.alloc(32, 0)` for clarity

File: lib/index.js (hunk 2 above)
Comment (fanatid) [inline lib/index.js]: @axic `new BN(i).toArrayLike(Buffer, 'be', 20)`?

Comment (axic) [inline lib/index.js]: Yep, 20 :)

Comment (axic) [inline lib/index.js]: I don't think `toArrayLike` uses the new Buffer API though.

File: package.json (hunk 4 above)
Comment (julianduque) [inline package.json]: @axic the browserify example complained so I needed to add this transformation, I can double check after if still needed

Comment (julianduque) [inline package.json]:
```

**Comment by julianduque** [comment]
@axic @fanatid implemented your suggestions and rebased again

File: package.json (hunk 4 above)
Comment (axic) [inline package.json]: Ah, so `stage-n` enables different set of features: https://stackoverflow.com/questions/37251552/whats-the-difference-between-babel-preset-stage-0-babel-preset-stage-1-etc

File: lib/runBlock.js [hunk 6]
```diff
@@ -84,7 +85,7 @@ module.exports = function (opts, cb) {
     function processTx (tx, cb) {
       var gasLimitIsHigherThanBlock = new BN(block.header.gasLimit).cmp(new BN(tx.gasLimit).add(gasUsed)) === -1
       if (gasLimitIsHigherThanBlock) {
-        cb('tx has a higher gas limit than the block')
+        cb(new Error('tx has a higher gas limit than the block'))
```
Comment (axic) [inline lib/runBlock.js]: I think this is not related to this PR.

File: lib/runCall.js [hunk 7]
```diff
@@ -28,7 +29,7 @@ module.exports = function (opts, cb) {
   var toAccount
   var toAddress = opts.to
   var createdAddress
-  var txValue = opts.value || new Buffer(0)
+  var txValue = opts.value || Buffer.alloc(0)
```
Comment (axic) [inline lib/runCall.js]: Please use `Buffer.from([0])` for clarity of intent.

Comment (julianduque) [inline lib/runCall.js]: In this case would be `Buffer.from([])`

File: lib/runBlock.js (hunk 6 above)
Comment (julianduque) [inline lib/runBlock.js]: oh, right, just a best practice, let me know if you want me to remove it from this PR.

File: package.json (hunk 4 above)
Comment (julianduque) [inline package.json]: @axic this is only to build the bundle in the examples, I can build the bundle and not include the dev dependency in `package.json`, since it's examples I think a new dependency isn't needed, what do you say?

File: lib/runCall.js (hunk 7 above)
Comment (julianduque) [inline lib/runCall.js]:
```

Comment (axic) [inline lib/runCall.js]: You see, this `alloc` feature is just confusing 😉 

I think in this case though both are acceptable as value can be zero or empty (still meaning zero).

Eventually all the values should be kept as BN and not as a Buffer.

File: lib/runBlock.js (hunk 6 above)
Comment (axic) [inline lib/runBlock.js]: I think it all of the callback errors should be of the Error class, but prefer to do it in a separate PR - there might be other cases too.

Comment (julianduque) [inline lib/runBlock.js]: @axic perfect, will remove this change from this PR and submit a new one with the proper errors.

**Comment by julianduque** [comment]
@axic done and done!, anything else?

File: lib/index.js [hunk 8]
```diff
@@ -46,7 +49,7 @@ function VM (opts = {}) {
 
   if (this.opts.activatePrecompiles) {
     for (var i = 1; i <= 4; i++) {
-      this.trie.put(new Buffer('000000000000000000000000000000000000000' + i, 'hex'), new Account().serialize())
+      this.trie.put(Buffer.from(new BN(i).toArray('be', 32, 20)), new Account().serialize())
```
Comment (axic) [inline lib/index.js]: I think this is wrong, should be `toArray('be', 20)`

**Comment by axic** [comment]
@julianduque can also please rebase (`git rebase origin/master`) instead of the merge commit?

It is good to go from my side after that.

**Comment by julianduque** [comment]
@axic thanks for the great feedback and review!

**Comment by axic** [comment]
Pulling the plug :)
